pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go

```diff
@@ -238,6 +238,59 @@ func (t ResolvedPipelineTask) hasRemainingRetries() bool {
 	return retriesDone < t.PipelineTask.Retries
 }
 
+// isCancelledForTimeOut returns true only if the run is cancelled due to PipelineRun-controlled timeout
+// If the PipelineTask has a Matrix, isCancelled returns true if any run is cancelled due to PipelineRun-controlled timeout and all other runs are done.
+func (t ResolvedPipelineTask) isCancelledForTimeOut() bool {
+	switch {
+	case t.IsCustomTask() && t.IsMatrixed():
+		if len(t.Runs) == 0 {
+			return false
+		}
+		isDone := true
+		atLeastOneCancelled := false
+		for _, run := range t.Runs {
+			isDone = isDone && run.IsDone()
+			c := run.Status.GetCondition(apis.ConditionSucceeded)
+			runCancelled := c.IsFalse() &&
+				c.Reason == v1alpha1.RunReasonCancelled &&
+				run.Spec.StatusMessage == v1alpha1.RunCancelledByPipelineTimeoutMsg
```

Comment:
I don't love having to do this, but we needed some way to determine that a `Run` or `TaskRun` that was nominally cancelled due to the `PipelineRun` deciding it had timed out was cancelled for that reason and marked as a failure, for consistency with how we behave if a `TaskRun`/`Run` with an explicit timeout set on its spec times out. The easiest way to do that was to look at `.Spec.StatusMessage`, which will be set on any `Run` or `TaskRun` we cancel-for-time-out.

Reply:
I'm not sure I understand the reason for this-- I only see this function used in pipelinerunstate.go to increment the count of failed tasks. why is this necessary?

Reply:
Because otherwise a `Run` or `TaskRun` that gets cancelled-for-timeout will show up in `facts.getPipelineTasksCount()` as cancelled, and [the `PipelineRun` itself will be marked as cancelled](https://github.com/tektoncd/pipeline/blob/bd8958cb09e1f75f95c7d7edce3e4b54f5c861f9/pkg/reconciler/pipelinerun/resources/pipelinerunstate.go#L505-L509).

Reply:
It seems weird that we decide whether the PipelineRun is canceled based on whether any of its child taskruns are canceled; I would think it should be the other way around. can we change this?

Reply:
Since that's a behavioral change outside of the scope of the particular issue I'm trying to fix here, I'd prefer to deal with that separately.

Reply:
so I understand, the current behavior is that both:
1. when a pipelinerun is cancelled, all its taskruns are cancelled
2. when a taskrun is cancelled, the parent pipelinerun is also cancelled

is that correct @abayer @lbernick?

Reply:
Yeah, the `PipelineRun` reconciler will mark the `PipelineRun` as cancelled if any of the `TaskRun`s or `Run`s in it are marked as cancelled.

Reply:
This sounds like a very weird behavior, but I agree, if we want to change it, it shouldn't be in that PR.